hbase-rsgroup/src/main/java/org/apache/hadoop/hbase/rsgroup/RSGroupAdminServer.java

```diff
@@ -198,108 +200,99 @@ private void checkServersAndTables(Set<Address> servers, Set<TableName> tables,
-          for (RegionInfo region : getRegions((Address) element)) {
-            if (!targetGrp.containsTable(region.getTable())) {
-              toMoveRegions.add(region);
-            }
-          }
-        }
-
-        //move regions
-        for (RegionInfo region : toMoveRegions) {
-          LOG.info("Moving server region {}, which do not belong to RSGroup {}",
+      for (Iterator<T> iter = allOwners.iterator(); iter.hasNext();) {
+        T owner = iter.next();
+        // Get regions that are associated with this server and filter regions by group tables.
+        for (RegionInfo region : getRegionsInfo.apply(owner)) {
+          if (!validation.apply(region)) {
+            LOG.info("Moving region {}, which do not belong to RSGroup {}",
               region.getShortNameToLog(), targetGroupName);
-          try {
-            this.master.getAssignmentManager().move(region);
-            failedRegions.remove(region.getRegionNameAsString());
-          } catch (IOException ioe) {
-            LOG.debug("Move region {} from group failed, will retry, current retry time is {}",
+            try {
+              this.master.getAssignmentManager().move(region);
+            }catch (IOException ioe){
+              LOG.error("Move region {} from group failed, will retry, current retry time is {}",
                 region.getShortNameToLog(), retry, ioe);
-            toThrow = ioe;
-            failedRegions.add(region.getRegionNameAsString());
-          }
-          if (master.getAssignmentManager().getRegionStates().
+            }
+            if (master.getAssignmentManager().getRegionStates().
               getRegionState(region).isFailedOpen()) {
-            continue;
+              continue;
+            }
+            hasRegionsToMove = true;
           }
-          hasRegionsToMove = true;
         }
+
         if (!hasRegionsToMove) {
-          LOG.info("There are no more regions of {} to move for RSGroup {}", element,
-              targetGroupName);
+          LOG.info("No more regions to move from {} to RSGroup", getOwnerName.apply(owner));
           iter.remove();
         }
       }
+
       retry++;
       try {
         rsGroupInfoManager.wait(1000);
       } catch (InterruptedException e) {
         LOG.warn("Sleep interrupted", e);
         Thread.currentThread().interrupt();
       }
-    } while (hasRegionsToMove && retry <= moveMaxRetry);
-
-    //has up to max retry time or there are no more regions to move
-    if (hasRegionsToMove) {
-      // print failed moved regions, for later process conveniently
-      String msg = String.format("move regions for group %s failed, failed regions: %s",
-          targetGroupName, failedRegions);
-      LOG.error(msg);
-      throw new DoNotRetryIOException(msg +
-          ", just record the last failed region's cause, more details in server log", toThrow);
-    }
+    } while (hasRegionsToMove && retry <= 50);
```

Comment:
50 can be replaced by moveMaxRetry.
And if the number of retries has reached the limit but there are still some regions need to be moved, should we throw out an Exception to notify the failure?

Reply:
> 50 can be replaced by 'moveMaxRetry'.

Definitely, missed that somehow when refactoring original method into this skeleton.

> If the number of retries has reached the limit but there are still some regions need to be moved, should we throw out an Exception to notify the failure?

Well spotted also. Yeah, I think if line #268 keeps throwing exception on all retries, we should keep original behaviour and throw exception upward. Maybe use a set to track which regions are failing, then mentioned those on the exception message? Also, I think we should change log verbosity on line #270 to debug, as it would make the log horrible verbose in the event of move failures even after 50 retries. 

@sunhelly , can you work on these improvements, then push another commit to this current PR?

Reply:
Yes, thanks your comments.